internal/storage/fs/git/cache.go

```diff
@@ -0,0 +1,161 @@
+package git
+
+import (
+	"context"
+	"sync"
+
+	lru "github.com/hashicorp/golang-lru/v2"
+	"golang.org/x/exp/maps"
+)
+
+type buildFunc[K comparable, V any] func(context.Context, K) (V, error)
+
+// cache contains a fixed set of non-evictable entries along
+// with additional capacity stored in an LRU.
+// The cache is keyed by reference (type R) with an indirect index
+// (type K for key) through into the stored values (type V)
+// The cache is provided with a resolver on creation, which is
+// used to dynamically resolve the key (K) from provided (R).
+type cache[R, K comparable, V any] struct {
+	mu sync.RWMutex
+
+	fixed map[R]K
+	extra *lru.Cache[R, K]
+
+	store map[K]V
+}
+
+func newCache[R, K comparable, V any](extra int) (_ *cache[R, K, V], err error) {
+	c := &cache[R, K, V]{
+		fixed: map[R]K{},
+	}
+
+	c.extra, err = lru.NewWithEvict[R, K](extra, c.evict)
+	if err != nil {
+		return nil, err
+	}
+
+	return c, nil
+}
+
+// AddFixed forcibly adds the revision, key and value tuple to fixed storage.
+// The supplied revision will never be evicted.
+// Subsequent calls to Add with the save value for r will update the fixed entries.
+func (c *cache[R, K, V]) AddFixed(ctx context.Context, r R, k K, v V) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	c.fixed[r] = k
+	c.store[k] = v
+
+	return
+}
+
+// AddOrBuild adds the revision, key and value tuple.
+// If the reference r is already tracked in the fixed set, then it is updated there.
+// Otherwise, the entry is added to the LRU cache.
+func (c *cache[R, K, V]) AddOrBuild(ctx context.Context, r R, k K, build buildFunc[K, V]) (V, error) {
+	v, ok, err := c.getOrBuild(ctx, r, k, build)
+	if err != nil {
+		return v, err
+	}
+
+	// fast path: r, k and v are all as expected already
+	if ok {
+		return v, nil
+	}
+
+	// Otherwise, either r did not resolve to k or there was
+	// no v stored for k
+	// In all cases we update both atomically to ensure that
+	// we don't get a race condition leading to a broken
+	// reference to key to value link
+
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	if _, ok := c.fixed[r]; ok {
+		c.fixed[r] = k
+	} else {
+		c.extra.Add(r, k)
+	}
+	c.store[k] = v
+
+	return v, nil
+}
+
+// Get attempts to resolve the value V for a given reference r.
+func (c *cache[R, K, V]) Get(r R) (v V, ok bool) {
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+
+	k, ok := c.fixed[r]
+	if ok {
+		v, ok = c.store[k]
+		return
+	}
+
+	k, ok = c.extra.Get(r)
+	if !ok {
+		return v, ok
+	}
+
+	v, ok = c.store[k]
+	return
+}
+
+func (c *cache[R, K, V]) getOrBuild(ctx context.Context, r R, k K, build buildFunc[K, V]) (v V, ok bool, err error) {
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+
+	v, ok = c.store[k]
+	if !ok {
+		// if there is no V for the provided k then we
+		// return early with ok is false to let the caller
+		// know to update all the references
+		v, err = build(ctx, k)
```

Comment:
Note to self: in hindsight I might be able to move this out if I have a sentinel way to say to the caller they have to build it themselves.